api/censuses.go

```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
+// the census merkle tree on IPFS. Then saves the resulting information of the
+// census tree in the database.
+func (capi *census3API) createAndPublishCensus(req *CreateCensusResquest, qID string) (int, error) {
+	bgCtx, cancel := context.WithTimeout(context.Background(), censusCreationTimeout)
+	defer cancel()
 	// begin a transaction for group sql queries
-	tx, err := capi.db.BeginTx(internalCtx, nil)
+	tx, err := capi.db.BeginTx(bgCtx, nil)
 	if err != nil {
-		return ErrCantCreateCensus
+		log.Errorw(err, "error starting database")
+		return -1, ErrCantCreateCensus.With("error starting database")
 	}
 	defer func() {
-		if err := tx.Rollback(); err != nil {
+		if err := tx.Rollback(); err != nil && !errors.Is(sql.ErrTxDone, err) {
 			log.Errorw(err, "holders transaction rollback failed")
 		}
 	}()
 	qtx := capi.sqlc.WithTx(tx)
-
-	strategyTokens, err := qtx.TokensByStrategyID(internalCtx, int64(req.StrategyID))
+	// get the tokens of the strategy provided and check them
+	strategyTokens, err := qtx.TokensByStrategyID(bgCtx, int64(req.StrategyID))
 	if err != nil {
 		if errors.Is(sql.ErrNoRows, err) {
 			log.Errorf("no strategy found for id %d: %s", req.StrategyID, err.Error())
-			return ErrNotFoundStrategy
+			return -1, ErrNoStrategyTokens.With("no strategy found")
 		}
 		log.Errorf("error getting strategy with id %d: %s", req.StrategyID, err.Error())
-		return ErrCantGetStrategy
+		return -1, ErrCantCreateCensus.With("error getting strategy")
+	}
+	if len(strategyTokens) == 0 {
+		log.Errorf("no tokens for strategy %d", req.StrategyID)
+		return -1, ErrNoStrategyTokens.With("no tokens for strategy")
+	}
+
+	// get the maximun current census ID to calculate the next one, if any
+	// census has been created yet, continue
+	lastCensusID, err := qtx.LastCensusID(bgCtx)
+	if err != nil && !errors.Is(sql.ErrNoRows, err) {
+		log.Errorw(err, "error getting last census ID")
+		return -1, ErrCantCreateCensus.With("error getting last census ID")
+	}
+	// compute the new censusId and censusType
+	newCensusID := int(lastCensusID) + 1
+	censusType := census.DefaultCensusType
+	if req.Anonymous {
+		censusType = census.AnonymousCensusType
 	}
 	// get holders associated to every strategy token, create a map to avoid
 	// duplicates and count the sum of the balances to get the weight of the
 	// census
 	censusWeight := new(big.Int)
 	strategyHolders := map[common.Address]*big.Int{}
 	for _, token := range strategyTokens {
-		holders, err := qtx.TokenHoldersByTokenID(internalCtx, token.ID)
+		holders, err := qtx.TokenHoldersByTokenID(bgCtx, token.ID)
 		if err != nil {
 			if errors.Is(sql.ErrNoRows, err) {
 				continue
 			}
-			log.Errorw(err, "error getting token holders")
-			return ErrCantGetTokenHolders.Withf("for the token with address %s",
-				common.BytesToAddress(token.ID))
+			log.Errorf("error getting token holders of %s: %v", common.BytesToAddress(token.ID), err)
```

Comment:
Don't print the error if you return it

Reply:
But probably the original error trace is useful for finding the root cause and at the same time, but does not provide any relevant information for the user of the API. This is why I usually print the error and return an "error response".

Reply:
Why not returning the full error (with details) to the API? My concern is that we are treating twice, in a different manner, the same error.